### DISCUSSION TITLE
CCQ Predoc Questions about FFT in the IPT Method (and CT-QMC)

### DISCUSSION THREAD

**Discussion by andrewkhardy** [TRIQS/triqs#894] [General]
Hello. I'm an incoming CCQ predoctoral student and I've been trying to go through the ModelDMFT tutorials. I apologize, but I have some simple questions about behind the scenes of TRIQS. 

The first of which is in the IPT tutorial with the Solver defined as

```
from triqs.gf import *
import numpy as np
from math import pi
class IPTSolver:
    def __init__(self, beta):
        self.beta = beta

        # Matsubara frequency Green's functions
        iw_mesh = MeshImFreq(beta=beta, S='Fermion', n_iw=1001)
        self.G_iw = Gf(mesh=iw_mesh, target_shape=[1,1])
        self.G0_iw = self.G_iw.copy() # self.G0 will be set by the user after initialization
        self.Sigma_iw = self.G_iw.copy()
        
        # Imaginary time
        tau_mesh = MeshImTime(beta=beta, S='Fermion', n_tau=10001)
        self.G0_tau = Gf(mesh=tau_mesh, target_shape=[1,1])
        self.Sigma_tau = self.G0_tau.copy()
        
    def solve(self, U):
        self.G0_tau << Fourier(self.G0_iw)
        self.Sigma_tau << (U**2) * self.G0_tau * self.G0_tau * self.G0_tau
        self.Sigma_iw << Fourier(self.Sigma_tau)
        
        # Dyson
        self.G_iw << inverse(inverse(self.G0_iw) - self.Sigma_iw)
```

I don't understand how we're able to Fourier Transform between imaginary time and Matsubara frequencies when the meshes are different sizes? What's the actual protocol going on here? I also have been taught to do FFT in factors of 2^n. Why do we instead want +1 odd numbered grids here?

**Comment by the-hampel** [comment]
Hi Andrew, 
some basic notations of the Fourier transform in triqs are explained here:
https://triqs.github.io/triqs/unstable/documentation/manual/triqs/gfs/basic_notions/contents.html
with more detailed implementation notes here: 
https://triqs.github.io/triqs/unstable/documentation/manual/triqs/gfs/c++/fourier_impl_notes.html
and an example here:
https://triqs.github.io/triqs/unstable/userguide/python/tutorials/Basics/01-Greens_functions.html#Fourier-transforms

But for your specific questions:
* yes indeed we want to have same number of frequencies and tau mesh for the final call to FFTW to perform the FT
* however triqs suggest to have 6 times more imaginary time (tau) points than Matsubara frequencies (w), this is an empirically well working ratio. 
* This is required because on the Gf(w) a tail fit (high frequency moments) will first be done and subtracted to make the Gf rapidly decaying, thus easier to FT. Then the vector of the data handed to FFTW will be extended from this tail fit or padded with zeros to match the number of tau points
* the +1 odd comes from the fact that we have to remove on the tau mesh the first / last point since it holds the same information (at least if I recall correctly)

If anyone has further insights on the current triqs implementation, please add / correct my points. :-) This is all I remember.

Best,
Alex
